README.md
# culqi_flutter

<a href="https://github.com/gucodero/culqi_flutter">
    <img src="https://i.ibb.co/R4ZN8Yz/culqi-flutter-logo.png"/>
</a>

## Resumen

Es una librería para flutter que implementa la creación de token en la pasarela de pagos Culqi.
Este repositorio no es oficial, por lo que no es mantenida por el equipo de desarrollo de Culqi.

## Características

* Proporciona una clase que gestiona los datos de la tarjeta.
* Brinda diferentes validaciones para evitar problemas innecesarios al enviar los datos bancarios a los servidores de Culqi.
* Implementa las excepciones arrojadas por la API de Culqi.
* Proporciona una clase con los datos generados por la API de Culqi al generar el token.

## Uso

Agregue esta línea a su archivo pubspec.yaml

```yaml

    dependencies:
        culqi_flutter: ^1.0.0

```

Importe el paquete de esta manera

```dart

    import 'package:culqi_flutter/culqi_flutter.dart';

```

Primero deberá pedir al usuario los datos de su tarjeta mediante un formulario para luego instanciar un objeto de la clase "CCard"

```dart

    CCard card = CCard(
      cardNumber: '4111 1111 1111 1111',
      expirationMonth: 09,
      expirationYear: 20,
      cvv: '123',
      email: 'test@test.com'
    );

```

Recuerda que la clase CCard tiene métodos que te ayudarán a validar los datos antes de enviarlo a la API de Culqi: 

* isCardNumberValid() return bool
* isYearValid() return bool
* isMonthValid() return bool
* isExpirationDateValid() return bool
* isCcvValid() return bool
* isEmailValid() return bool
* getBrand() return int - ejemplo

```dart

    Switch(card.getBrand()){
        case CCard.DINERS_CLUB:
            break;
        case CCard.AMERICAN_EXPRESS:
            break;
        case CCard.MASTER_CARD:
            break;
        case CCard.VISA:
            break;
        default:
            // no se detecta la marca de la tarjeta
            break;
    }

```

Para enviar los datos de la tarjeta hacemos uso del método asíncrono "createToken":

```dart

    createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59").then((CToken token){
        //su token
        print(token.id);
    }).catchError((error){
        try{
            throw error;
        } on CulqiBadRequestException catch(ex){
            print(ex.cause);
        } on CulqiUnknownException catch(ex){
            //codigo de error del servidor
            print(ex.cause);
        }
    });

```

Si utiliza aync/await su código se vería de la siguiente manera:

```dart

    try{
        CToken token = await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59");
        //su token
        print(token.id);
    } on CulqiBadRequestException catch(ex){
        print(ex.cause);
    } on CulqiUnknownException catch(ex){
        //codigo de error del servidor
        print(ex.cause);
    }

```

## LICENCIA

```

MIT License

Copyright (c) 2020 Gucodero

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

```

CHANGELOG.md
## [1.0.0]

* initRelease
* Clase CCard para validación de tarjeta de manera local
* Clase CToken con datos recibidos por los servidores de culqi al generar token
* Implementación de errores
* Metodo createToken para gestionar petición de token a servidor culqi
* Implementación de pruebas unitarias a CCard y createToken

test/culqi_flutter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:culqi_flutter/culqi_flutter.dart';

void main() {

  test('Is card number valid', () {
    CCard card = CCard(cardNumber: '4557 880 ');
    expect(card.isCardNumberValid(), false);
    card.cardNumber = '4557 8805 8520';
    expect(card.isCardNumberValid(), false);
    card.cardNumber = '4557 8805 8520 8147';
    expect(card.isCardNumberValid(), true);
    card.cardNumber = '4557 8805 8520 8146';
    expect(card.isCardNumberValid(), false);
    card.cardNumber = '4557 8805 852A 8146';
    expect(card.isCardNumberValid(), false);
  });

  test('Is year valid', (){
    CCard card = CCard(expirationYear: 10);
    expect(card.isYearValid(), false);
    card.expirationYear = 1;
    expect(card.isYearValid(), false);
    card.expirationYear = 19;
    expect(card.isYearValid(), false);
    card.expirationYear = 123;
    expect(card.isYearValid(), false);
    card.expirationYear = 20;
    expect(card.isYearValid(), true);
    card.expirationYear = 22;
    expect(card.isYearValid(), true);
  });

  test('Is Month valid', (){
    CCard card = CCard(expirationMonth: 0);
    expect(card.isMonthValid(), false);
    card.expirationMonth = 1;
    expect(card.isMonthValid(), true);
    card.expirationMonth = 12;
    expect(card.isMonthValid(), true);
    card.expirationMonth = 13;
    expect(card.isMonthValid(), false);
    card.expirationMonth = 5;
    expect(card.isMonthValid(), true);
  });

  test('Is expiration date valid', (){
    CCard card = CCard(expirationYear: 0, expirationMonth: 0);
    expect(card.isExpirationDateValid(), false);
    card.expirationMonth = DateTime.now().month+1;
    card.expirationYear = DateTime.now().year - 2000;
    expect(card.isExpirationDateValid(), true);
    card.expirationMonth = 12;
    card.expirationYear = DateTime.now().year - 2001;
    expect(card.isExpirationDateValid(), false);
    card.expirationMonth = DateTime.now().month;
    card.expirationYear = DateTime.now().year - 1999;
    expect(card.isExpirationDateValid(), true);
    card.expirationMonth = DateTime.now().month;
    card.expirationYear = DateTime.now().year;
    expect(card.isExpirationDateValid(), false);
  });

  test('Is cvv valid', () {
    CCard card = CCard(cvv: 'as');
    expect(card.isCcvValid(), false);
    card.cvv = '';
    expect(card.isCcvValid(), false);
    card.cvv = '1';
    expect(card.isCcvValid(), false);
    card.cvv = 'a23';
    expect(card.isCcvValid(), false);
    card.cvv = '13';
    expect(card.isCcvValid(), false);
    card.cvv = '123';
    expect(card.isCcvValid(), true);
    card.cvv = '1234';
    expect(card.isCcvValid(), true);
    card.cvv = '12345';
    expect(card.isCcvValid(), false);
  });

  test('Get brand', () {
    CCard card = CCard(cardNumber: '3000');
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '305';
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '302';
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '309';
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '36047';
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '38047';
    expect(card.getBrand(), CCard.DINERS_CLUB);
    card.cardNumber = '39';
    expect(card.getBrand(), CCard.DINERS_CLUB);  
    card.cardNumber = '34512';
    expect(card.getBrand(), CCard.AMERICAN_EXPRESS);
    card.cardNumber = '3712';
    expect(card.getBrand(), CCard.AMERICAN_EXPRESS);
    card.cardNumber = '5152';
    expect(card.getBrand(), CCard.MASTER_CARD);
    card.cardNumber = '5523';
    expect(card.getBrand(), CCard.MASTER_CARD);
    card.cardNumber = '45578805';
    expect(card.getBrand(), CCard.VISA);
    card.cardNumber = '4';
    expect(card.getBrand(), CCard.VISA);
    card.cardNumber = '128';
    expect(card.getBrand(), CCard.ANY);
  });

  test('Create Token', () async {
    CCard card = CCard(
      cardNumber: '4111 1111 1111 1111',
      expirationMonth: 09,
      expirationYear: 20,
      cvv: '123',
      email: 'pablo@hotmail.com'
    );
    expect(await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), isInstanceOf<CToken>());
    card = CCard(
      cardNumber: '4111 1111 1111 1111',
      expirationMonth: 09,
      expirationYear: 20,
      cvv: '123',
    );
    expect(() async => await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), throwsA(isInstanceOf<CulqiBadRequestException>()));
    card = CCard(
      cardNumber: '4111 1111 1111',
      expirationMonth: 09,
      expirationYear: 20,
      cvv: '123',
      email: 'pablo@hotmail.com'
    );
    expect(() async => await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), throwsA(isInstanceOf<CulqiBadRequestException>()));
    card = CCard(
      cardNumber: '4000 0200 0000 0000',
      expirationMonth: 10,
      expirationYear: 19,
      cvv: '354',
      email: 'pablo@hotmail.com'
    );
    expect(() async => await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), throwsA(isInstanceOf<CulqiBadRequestException>()));
    card = CCard(
      cardNumber: '4000 0300 0000 0009',
      expirationMonth: 08,
      expirationYear: 24,
      cvv: '836',
      email: 'pablo@hotmail.com'
    );
    expect(await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), isInstanceOf<CToken>());
    card = CCard(
      cardNumber: '4000 0400 0000 0008',
      expirationMonth: 03,
      expirationYear: 21,
      cvv: '295',
      email: 'pablo@hotmail.com'
    );
    expect(await createToken(card: card, apiKey: "sk_test_UTCQSGcXW8bCyU59"), isInstanceOf<CToken>());
  });
     
}

lib/src/CToken.dart
import 'dart:convert';

class CToken{

  String id;
  String type;
  DateTime creationDate;
  String email;
  String cardNumber;
  String lastFour;
  bool active;
  Iin iin;
  Client client;

  CToken.fromJson(String json){
    Map<String, dynamic> data = jsonDecode(json);
    id = data['id'];
    type = data['type'];
    creationDate = new DateTime.fromMillisecondsSinceEpoch(data['creation_date']);
    email = data['email'];
    cardNumber = data['card_number'];
    lastFour = data['last_four'];
    active = data['active'];
    iin = Iin.fromMap(data['iin']);
    client = Client.fromMap(data['client']);
  }

  Map<String, dynamic> toJson() => {
    'id' : '$id',
    'type' : '$type',
    'creationDate' : '$creationDate',
    'email' : '$email',
    'cardNumber' : '$cardNumber',
    'lastFour' : '$lastFour',
    'active' : '$active',
    'iin' : '${iin.toString()}',
    'client' : '${client.toString()}'
  };

  @override
  String toString() {
    return toJson().toString();
  }

}

class Iin {

  String bin;
  String cardBrand;
  String cardType;
  String cardCategory;
  List installmentsAllowed;
  Issuer issuer;

  Iin.fromMap(Map<String, dynamic> data){
    bin = data['bin'];
    cardBrand = data['card_brand'];
    cardType = data['card_type'];
    cardCategory = data['card_category'];
    installmentsAllowed = data['installments_allowed'];
    issuer = Issuer.fromMap(data['issuer']);
  }

  Map<String, dynamic> toJson() => {
    'bin' : '$bin',
    'cardBrand' : '$cardBrand',
    'cardType' : '$cardType',
    'cardCategory' : '$cardCategory',
    'installmentsAllowed' : '$installmentsAllowed',
    'issuer' : '${issuer.toString()}'
  };

  @override
  String toString() {
    return toJson().toString();
  }

}
class Issuer {

  String name;
  String country;
  String countryCode;
  String website;
  String phoneNumber;

  Issuer.fromMap(Map<String, dynamic> data){
    name = data['name'];
    country = data['country'];
    countryCode = data['country_code'];
    website = data['website'];
    phoneNumber = data['phone_number'];
  }

  Map<String, dynamic> toJson() => {
    'name' : '$name',
    'country' : '$country',
    'countryCode' : '$countryCode',
    'website' : '$website',
    'phoneNumber' : '$phoneNumber'
  };

  @override
  String toString() {
    return toJson().toString();
  }

}
class Client {

  String ip;
  String ipCountry;
  String ipCountryCode;
  String browser;
  String deviceFingerprint;
  String deviceType;

  Client.fromMap(Map<String, dynamic> data){
    ip = data['ip'];
    ipCountry = data['ip_country'];
    ipCountryCode = data['ip_country_code'];
    browser = data['browser'];
    deviceFingerprint = data['device_fingerprint'];
    deviceType = data['device_type'];
  }

  Map<String, dynamic> toJson() => {
    'ip' : '$ip',
    'ipCountry' : '$ipCountry',
    'ipCountryCode' : '$ipCountryCode',
    'browser' : '$browser',
    'deviceFingerprint' : '$deviceFingerprint',
    'deviceType' : '$deviceType'
  };

  @override
  String toString() {
    return toJson().toString();
  }

}


lib/culqi_flutter.dart
library culqi_flutter;

export 'src/CCard.dart';
export 'src/CToken.dart';
export 'src/errors/CulqiBadRequestException.dart';
export 'src/errors/CulqiUnknownException.dart';

import 'dart:convert';
import 'package:flutter/widgets.dart';
import 'package:http/http.dart' as http;
import 'package:culqi_flutter/src/CCard.dart';
import 'package:culqi_flutter/src/CToken.dart';
import 'package:culqi_flutter/src/errors/CulqiBadRequestException.dart';
import 'package:culqi_flutter/src/errors/CulqiUnknownException.dart';



const String _URL_BASE_SECURE = 'https://secure.culqi.com/v2';
const String _URL_TOKENS = '/tokens/';

Future<CToken> createToken({@required CCard card,@required String apiKey}) async{
    http.Response response = await http.post("$_URL_BASE_SECURE$_URL_TOKENS", headers: {
      'Content-Type': 'application/json; charset=utf-8',
      'Authorization': "Bearer $apiKey"
    }, body: jsonEncode(card.toJson()));
    switch(response.statusCode){
      case 201:
        return CToken.fromJson(response.body);
      case 400:
        throw CulqiBadRequestException.fromJson(response.body);
      default: 
        throw CulqiUnknownException(response.statusCode.toString(), response.body);
    }
}





lib/src/CCard.dart
import 'package:email_validator/email_validator.dart';

class CCard {

  static const int ANY = 0, VISA = 1, MASTER_CARD = 2, 
  DINERS_CLUB = 4, AMERICAN_EXPRESS = 5; 

  String _cardNumber;
  String cvv;
  int expirationMonth;
  int expirationYear;
  String email;

  CCard({ String cardNumber = '', this.cvv = '', this.expirationMonth = 0, this.expirationYear = 0, this.email = ''}){
    this._cardNumber = cardNumber.replaceAll(' ', '');
  }
  
  String get cardNumber {
    return _cardNumber;
  }

  set cardNumber(String value){
    _cardNumber = value.replaceAll(' ', '');
  }

  //luhn algorithm
  bool isCardNumberValid(){
    int size = _cardNumber.length;
    //invertimos el numero de tarjeta
    String cardNumberReverse = String.fromCharCodes(_cardNumber.runes.toList().reversed);
    try{
      int.parse(cardNumber);
    }catch(_){
      return false;
    }
    // Sin necesidad de aplicar algoritmo
    if(size < 8){
      return false;
    }
    
    int sum = 0;
    //recorremos digito por digito
    for(int i = 0; i < size; i++){
      int digit = int.parse(cardNumberReverse[i]); 
      //si la posicion es impar contando desde 0 duplicamos el valor
      if(i % 2 == 1){
        digit *= 2;
      }

      //si el numero tiene dos digitos le restamos 9 para que vuelva a tener un digito

      sum += digit > 9 ? (digit - 9) : digit;

    }
    return sum % 10 == 0;
  }

  bool isYearValid(){
    if(expirationYear > 99){
      return false;
    }
    return (expirationYear+2000) >= DateTime.now().year;
  }

  bool isMonthValid(){
    return expirationMonth <= 12 && expirationMonth >= 1;
  }

  bool isExpirationDateValid(){
    if(isYearValid() && isMonthValid()){
      if((expirationYear+2000) == DateTime.now().year){
        return expirationMonth > DateTime.now().month;
      }
      return true;
    }
    return false;
  }

  bool isCcvValid(){

    try{
      int.parse(cvv);
    }catch(_){
      return false;
    }
    
    return cvv.length == 3 || cvv.length == 4;

  }
  
  bool isEmailValid(){
    return EmailValidator.validate(email);
  }

  int getBrand(){
    String card = _cardNumber.toString();
    
    if(card.length >= 3){
      int n = int.parse(card.substring(0,3));
      if((n >= 300 && n <= 305) || n == 309){
        return DINERS_CLUB;
      }
    }
    
    if(card.length >= 2){
      int n = int.parse(card.substring(0,2));
      if(n == 36 || n == 38 || n == 39){
        return DINERS_CLUB;
      }else if(n == 34 || n == 37){
        return AMERICAN_EXPRESS;
      }else if(n >= 51 && n <= 55){
        return MASTER_CARD;
      }
    }
    
    if(card.length >= 1){
      int n = int.parse(card[0]);
      if(n == 4){
        return VISA;
      }
    }
    return ANY;
  }

  Map<String, dynamic> toJson() => 
    {
      'card_number': '$cardNumber',
      'cvv': '$cvv',
      'expiration_month': expirationMonth,
      'expiration_year': expirationYear + 2000,
      'email': email
    };

}